plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs

```diff
@@ -0,0 +1,32 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE LambdaCase          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module UntypedPlutusCore.Check.Scope (checkScope) where
+
+import Control.Lens hiding (index)
+import UntypedPlutusCore.Core.Type as UPLC
+import UntypedPlutusCore.DeBruijn as UPLC
+
+checkScope :: forall name uni fun a. HasIndex name
+           => UPLC.Term name uni fun a
+           -> Bool
+checkScope = \case
+    -- there is no "debruijn-level" at the toplevel
+    Var _ _       -> False
+    LamAbs _ _ t  -> withL 1 t
+    Apply _ t1 t2 -> checkScope t1 && checkScope t2
+    Force _ t     -> checkScope t
+    Delay _ t     -> checkScope t
+    _             -> True
+  where
+    -- the current level as a reader value
+    withL :: Word -> UPLC.Term name uni fun a -> Bool
+    withL !lvl = \case
+        -- var should be larger than 0 (satisfied by Index=Natural/Word)
+        -- var should be LEQ to the current level
+        Var _ n       -> fromIntegral (n ^. index) <= lvl
+        LamAbs _ _ t  -> withL (lvl+1) t
```

Comment:
Hmmm. What happened in the past when the index was wrong (i.e. not 0)? I see that [now](https://github.com/input-output-hk/plutus/blob/master/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/DeBruijn.hs#L138) we ignore it and just act as if it was 0, but I think that's not what we did in the actual Alonzo HF, we actually treated the index as if it mattered (this is true, the behaviour changed [here](https://github.com/input-output-hk/plutus/commit/da4f85cdd2a3a261ce540e8dc51d2a3c5fa89ed2#diff-33151f0a151d48e22c6d4855933528c1c7fbace6dd470d218bc27ba38c25e121).

However, answering my own question: we omit indices on binders in the serialized form, and put them in as 0 when we deserialise, so we can never have wrong indices. Nonetheless, I think it's cheap and sensible to check that they're all 0 here. With a comment.

Reply:
fair enough.